src/imitation/regularization/updaters.py

```diff
@@ -0,0 +1,139 @@
+
+
+class IntervalParamScaler(LambdaUpdater):
+    """Scales the lambda of the regularizer by some constant factor.
+
+    Lambda is scaled up if the ratio of the validation loss to the training loss
+    is above the tolerable interval, and scaled down if the ratio is below the
+    tolerable interval. Nothing happens if the ratio is within the tolerable
+    interval.
+    """
+
+    def __init__(self, scaling_factor: float, tolerable_interval: Tuple[float, float]):
+        """Initialize the interval parameter scaler.
+
+        Args:
+            scaling_factor: The factor by which to scale the lambda, a value in (0, 1).
+            tolerable_interval: The interval within which the ratio of the validation
+                loss to the training loss is considered acceptable. A tuple whose first
+                element is at least 0 and the second element is greater than the first.
+
+        Raises:
+            ValueError: If the tolerable interval is not a tuple of length 2.
+            ValueError: if the scaling factor is not in (0, 1).
+            ValueError: if the tolerable interval is negative or not a proper interval.
+        """
+        eps = np.finfo(float).eps
+        if not (eps < scaling_factor < 1 - eps):
+            raise ValueError(
+                "scaling_factor must be in (0, 1) within machine precision.",
+            )
+        if len(tolerable_interval) != 2:
+            raise ValueError("tolerable_interval must be a tuple of length 2")
+        if not (0 <= tolerable_interval[0] < tolerable_interval[1]):
+            raise ValueError(
+                "tolerable_interval must be a tuple whose first element "
+                "is at least 0 and the second element is greater than "
+                "the first",
+            )
+
+        self.scaling_factor = scaling_factor
+        self.tolerable_interval = tolerable_interval
+
+    def __call__(
+        self,
+        lambda_: float,
+        train_loss: LossType,
+        val_loss: LossType,
+    ) -> float:
+        """Scales the lambda of the regularizer by some constant factor.
+
+        Lambda is scaled up if the ratio of the validation loss to the training loss
+        is above the tolerable interval, and scaled down if the ratio is below the
+        tolerable interval. Nothing happens if the ratio is within the tolerable
+        interval.
+
+        Args:
+            lambda_: The current value of the lambda.
+            train_loss: The loss on the training set.
+            val_loss: The loss on the validation set.
+
+        Returns:
+            The new value of the lambda.
+
+        Raises:
+            ValueError: If the loss on the validation set is not a scalar.
+            ValueError: if lambda_ is zero (will result in no scaling).
+            ValueError: if lambda_ is not a float.
+        """
+        # assert that the tensors val_loss and train_loss are both scalars
+        try:
+            assert isinstance(val_loss, float) or (
```

Comment:
Why the indirection: assert then trap it and raise ValueError? Note `assert` is disabled when Python runs in optimized mode `-O` so it's not just a distinction between `AssertionError` and `ValueError` (although optimized mode is so rarely used the importance of this is minor).

One alternative would be to define a helper function that does this input validation, and call it for both `train_loss` and `val_loss`

Reply:
Yeah you're right, I was not convinced by writing it this way, but somehow ended up leaving it. I guess I found it easier to write a bunch of checks and write only one error message, rather than split it up.